src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/TeredoHelper.cs

```diff
@@ -84,85 +65,67 @@ public static bool UnsafeNotifyStableUnicastIpAddressTable(Action<object> callba
-                if (Environment.HasShutdownStarted)
-                {
-                    return;
-                }
+            if (_gcHandle.IsAllocated)
+                _gcHandle.Free();
+        }
 
-#if DEBUG
-                bool successfullyRemoved = s_pendingNotifications.Remove(this);
-                if (!successfullyRemoved)
-                {
-                    NetEventSource.Fail(null, "RunCallback for a TeredoHelper which is not in s_pendingNotifications!");
-                }
-#else
-                s_pendingNotifications.Remove(this);
-#endif
+        private static void RunCallback(object? o)
+        {
+            TeredoHelper helper = (TeredoHelper)o!;
 
-                if ((_cancelHandle == null || _cancelHandle.IsInvalid))
-                {
-                    NetEventSource.Fail(null, "Invalid cancelHandle in RunCallback");
-                }
+            Debug.Assert(helper._runCallbackCalled);
+            Debug.Assert(helper._cancelHandle != null && !helper._cancelHandle.IsInvalid);
 
-                _cancelHandle.Dispose();
-            }
+            helper.Dispose();
 
-            _callback.Invoke(_state);
+            helper._callback.Invoke(helper._state);
         }
 
         // This callback gets run on a native worker thread, which we don't want to allow arbitrary user code to
         // execute on (it will block AppDomain unload, for one).  Free the MibTable and delegate (exactly once)
         // to the managed ThreadPool for the rest of the processing.
-        //
-        // We can't use SafeHandle here for table because the marshaller doesn't support them in reverse p/invokes.
-        // We won't get an AppDomain unload here anyways, because OnAppDomainUnloaded will block until all of these
-        // callbacks are done.
-        private void OnStabilized(IntPtr context, IntPtr table)
+        [UnmanagedCallersOnly]
+        private static void OnStabilized(IntPtr context, IntPtr table)
         {
             Interop.IpHlpApi.FreeMibTable(table);
 
+            TeredoHelper helper = (TeredoHelper)GCHandle.FromIntPtr(context).Target!;
+
             // Lock the TeredoHelper instance to ensure that only the first call to OnStabilized will get to call
             // RunCallback.  This is the only place that TeredoHelpers get locked, as individual instances are not
             // exposed to higher layers, so there's no chance for deadlock.
-            if (!_runCallbackCalled)
+            if (!helper._runCallbackCalled)
             {
-                lock (this)
+                lock (helper)
                 {
-                    if (!_runCallbackCalled)
+                    if (!helper._runCallbackCalled)
                     {
-                        _runCallbackCalled = true;
-                        ThreadPool.QueueUserWorkItem(RunCallback, null);
+                        helper._runCallbackCalled = true;
+                        ThreadPool.QueueUserWorkItem(RunCallback, helper);
```

Comment:
Seems like it'd be cleaner and cheaper to just inline the callback code:
```C#
ThreadPool.QueueUserWorkItem(o =>
{
    TeredoHelper helper = (TeredoHelper)o!;
    helper.Dispose();
    helper._callback.Invoke(helper._state);
}, helper);
```
and then delete RunCallback.